crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+ *
+ * Encrypts a message with given randomness to the ciphertext in |out|. Without
+ * applying the Fujisaki-Okamoto transform this would not result in a CCA
+ * secure scheme, since lattice schemes are vulnerable to decryption failure
+ * oracles.
+ *
+ * Caller passes storage for |y|, |e1| and |u|.
+ */
+static __owur
+int encrypt_cpa(uint8_t *out, const uint8_t *message, const scalar *m,
+                const scalar *t, const uint8_t *randomness, scalar *y,
+                scalar *e1, scalar *u, vinfo_t vinfo, mctx *ctx)
+{
+    uint8_t counter = 0;
+    uint8_t input[ML_KEM_RANDOM_BYTES + 1];
+    scalar e2;
+    scalar v;
+    scalar expanded_message;
+    int rank = vinfo->rank;
+
+    /* FIPS 203 "y" vector */
+    if (!gencbd_vector(y, vinfo->cbd1, &counter, randomness, rank, ctx))
+        return 0;
+    vector_ntt(y, rank);
+    /* FIPS 203 "e1" vector */
+    if (!gencbd_vector(e1, vinfo->cbd2, &counter, randomness, rank, ctx))
+        return 0;
+    memcpy(input, randomness, ML_KEM_RANDOM_BYTES);
+    input[ML_KEM_RANDOM_BYTES] = counter;
+    /* FIPS 203 "e2" scalar */
+    if (!vinfo->cbd2(&e2, input, ctx))
+        return 0;
+    /* FIPS 203 "u" vector */
+    matrix_mult(u, m, y, rank);
+    vector_inverse_ntt(u, rank);
+    vector_add(u, e1, rank);
+    /* FIPS 203 "v" scalar */
+    inner_product(&v, t, y, rank);
+    scalar_inverse_ntt(&v);
+    scalar_add(&v, &e2);
+    /* Extract ciphertext */
+    scalar_decode_1(&expanded_message, message);
+    scalar_decompress(&expanded_message, 1);
+    scalar_add(&v, &expanded_message);
+    vector_compress(u, vinfo->du, rank);
+    vector_encode(out, u, vinfo->du, rank);
+    scalar_compress(&v, vinfo->dv);
+    scalar_encode(out + vinfo->u_vector_bytes, &v, vinfo->dv);
+    return 1;
+}
+
+/*
+ * FIPS 203, Section 5.3, Algorithm 15: K-PKE.Decrypt.
+ */
+static void
+decrypt_cpa(uint8_t *out, const uint8_t *ctext, const scalar *s, scalar *u,
+            vinfo_t vinfo)
+{
+    scalar v, mask;
+    int rank = vinfo->rank;
+    int du = vinfo->du;
+    int dv = vinfo->dv;
+
+    vector_decode(u, ctext, du, rank);
+    vector_decompress(u, du, rank);
+    vector_ntt(u, rank);
+    scalar_decode(&v, ctext + vinfo->u_vector_bytes, dv);
+    scalar_decompress(&v, dv);
+    inner_product(&mask, s, u, rank);
+    scalar_inverse_ntt(&mask);
+    scalar_sub(&v, &mask);
+    scalar_compress(&v, 1);
+    scalar_encode_1(out, &v);
+}
+
+/*
+ * -----
+ *
+ * Internal API implements details of variant-specific ML-KEM algorithms for
+ * multiple variants given explicit parameters and additional caller allocated
+ * temporary storage as needed.
+ */
+
+mctx *ossl_ml_kem_newctx(OSSL_LIB_CTX *libctx, const char *properties)
+{
+    mctx *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    /* Precondition of ML-KEM implementation correctness */
+    CHECK_SIZE_UINT_GE_UINT32;
+
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
```

Comment:
CRYPTO_malloc() was changed a while ago to raise an ERR_R_MALLOC_FAILURE, so this is not needed I think

Reply:
Noted, the inherited instance of this will shortly be gone from #16172.